go/vt/discovery/healthcheck.go

```diff
@@ -0,0 +1,283 @@
+}
+
+// healthCheckConn contains details about an endpoint.
+type healthCheckConn struct {
+	// set at construction time
+	cell       string
+	cancelFunc context.CancelFunc
+
+	// mu protects all the following fields
+	mu     sync.RWMutex
+	conn   tabletconn.TabletConn
+	target *pbq.Target
+	stats  *pbq.RealtimeStats
+}
+
+// checkConn performs health checking on the given endpoint.
+func (hc *HealthCheck) checkConn(cell string, endPoint *pbt.EndPoint) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	hcc := &healthCheckConn{
+		cell:       cell,
+		cancelFunc: cancelFunc,
+	}
+
+	// retry health check if it fails
+	for {
+		stream, errfunc, err := hcc.connect(ctx, hc, endPoint)
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			log.Errorf("cannot connect to %+v: %v", endPoint, err)
+			time.Sleep(hc.retryDelay)
+			continue
+		}
+		for {
+			err = hcc.processResponse(ctx, hc, endPoint, stream, errfunc)
+			if err != nil {
+				hcc.mu.Lock()
+				hcc.conn.Close()
+				hcc.conn = nil
+				hcc.mu.Unlock()
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				log.Errorf("error when streaming tablet health from %+v: %v", endPoint, err)
+				time.Sleep(hc.retryDelay)
+				break
+			}
+		}
+	}
+}
+
+// connect creates connection to the endpoint and starts streaming.
+func (hcc *healthCheckConn) connect(ctx context.Context, hc *HealthCheck, endPoint *pbt.EndPoint) (<-chan *pbq.StreamHealthResponse, tabletconn.ErrFunc, error) {
+	conn, err := tabletconn.GetDialer()(ctx, endPoint, "" /*keyspace*/, "" /*shard*/, pbt.TabletType_RDONLY, hc.connTimeout)
+	if err != nil {
+		return nil, nil, err
+	}
+	stream, errfunc, err := conn.StreamHealth(ctx)
+	if err != nil {
+		conn.Close()
+		return nil, nil, err
+	}
+	hcc.mu.Lock()
+	hcc.conn = conn
+	hcc.mu.Unlock()
+	return stream, errfunc, nil
+}
+
+// processResponse reads one health check response, and notifies HealthCheckStatsListener.
+func (hcc *healthCheckConn) processResponse(ctx context.Context, hc *HealthCheck, endPoint *pbt.EndPoint, stream <-chan *pbq.StreamHealthResponse, errfunc tabletconn.ErrFunc) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case shr, ok := <-stream:
+		if !ok {
+			return errfunc()
+		}
+		if shr.Target == nil || shr.RealtimeStats == nil || shr.RealtimeStats.HealthError != "" {
+			return fmt.Errorf("health stats is not valid: %v", shr)
+		}
+
+		if hcc.target == nil {
+			// The first time we see response for the endpoint.
+			hcc.mu.Lock()
+			hcc.target = shr.Target
+			hcc.stats = shr.RealtimeStats
+			hcc.mu.Unlock()
+			hc.mu.Lock()
+			key := hc.endPointToMapKey(endPoint)
+			hc.addrToConns[key] = hcc
+			hc.addEndPointToTargetProtected(hcc.target, endPoint)
+			hc.mu.Unlock()
+		} else if hcc.target.TabletType != shr.Target.TabletType {
+			hc.mu.Lock()
+			hc.deleteEndPointFromTargetProtected(hcc.target, endPoint)
+			hcc.mu.Lock()
```

Comment:
Here you need to either document above (in the structure definitions) the order in which the locks are taken, or not take both locks at once. This makes it much clearer on how to avoid deadlocks.